File: c/io/fileManager/FileManager.cpp
#include "FileManager.h"

#include "cocos2d.h"
USING_NS_CC;

#if (CC_TARGET_PLATFORM != CC_PLATFORM_WIN32)
#include <dirent.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <errno.h>
#endif
using namespace std;


FileManager *FileManager::m_instance = NULL;


FileManager::FileManager()
{

}


FileManager::~FileManager()
{
    destroyInstance();
}


FileManager *FileManager::getInstance()
{
    if(!m_instance)
    {
        m_instance = new FileManager;
    }
    return m_instance;
}


void FileManager::destroyInstance()
{
    if (!m_instance)
    {
        delete m_instance;
        m_instance = NULL;
    }
}

string FileManager::convertPath(std::string path, int pos)
{
#if CC_TARGET_PLATFORM == CC_PLATFORM_WIN32
    int nPos = path.find("/", pos);
    if (nPos >= 0)
    {
        path.replace(nPos, 1, "\\");
        return convertPath(path, nPos + 1);
    }
    else
    {
        return path;
    }
#else
    int nPos = path.find("\\", pos);
    if (nPos >= 0)
    {
        path.replace(nPos, 1, "/");
        return convertPath(path, nPos + 1);
    }
    else
    {
        return path;
    }
#endif
}

bool FileManager::copyFile(std::string originPath, std::string newPath)
{    
    //检查文件是否存在
    if (!CCFileUtils::sharedFileUtils()->isFileExist(originPath))
    {
        CCLog("%s", ("File:" + originPath + " don't exist!").c_str());

        return false;
    }

    unsigned long len = 0;
    unsigned char *data = CCFileUtils::sharedFileUtils()->getFileData(originPath.c_str(), "rb", &len);

    //检查是否打开成功
    if (!data)
    {
        CCLog("%s", ("Get data from:" + originPath + " error!").c_str());

        return false;
    }

    //由于新文件的路径可能存在未创建的文件夹，所以需要便利创建文件夹才行
    int pos = newPath.find_last_of("\\/");
    int size = newPath.size();
    if (pos < 0 || pos >= size)
    {
        CCLog("%s", ("New path:" + originPath + " error!").c_str());

        return false;
    }

    //生成路径文件夹
    string path = newPath.substr(0, pos);
    if (!createFolderRecursively(path))
    {
        CCLog("%s", ("Can't create folder:" + path).c_str());

        return false;
    }
    
    FILE *file = fopen(newPath.c_str(), "wb");
    //检查是否打开成功
    if (!file)
    {
        newPath = CCFileUtils::sharedFileUtils()->fullPathForFilename(newPath.c_str());
        CCLog("%s", ("Can't open file:" + newPath).c_str());

        return false;
    }

    fwrite(data, sizeof(char), len, file);
    fclose(file);

    return true;
}


bool FileManager::reName(std::string originPath, std::string newPath)
{
    originPath = convertPath(originPath);
    newPath = convertPath(newPath);

    //获取文件名
    int pos = newPath.find_last_of("/\\");
    string newName = newPath.substr(pos + 1);
    //先删除原来的文件
    deleteFile(newPath);

    if (!system(("rename " + originPath + " " + newName).c_str()))
    {
        return true;
    }
    else
    {
        return false;
    }
}


bool FileManager::createFolder(std::string path)
{
    path = convertPath(path);

#if (CC_TARGET_PLATFORM != CC_PLATFORM_WIN32)
    mode_t processMask = umask(0);
    int ret = mkdir(path.c_str(), S_IRWXU | S_IRWXG | S_IRWXO);
    umask(processMask);
    if (ret != 0 && errno != EEXIST)
    {
        return false;
    }
    else
    {
        return true;
    }
#else
    BOOL ret = CreateDirectoryA(path.c_str(), NULL);
    if (!ret && ERROR_ALREADY_EXISTS != GetLastError())
    {
        return false;
    }
    else
    {
        return true;
    }
#endif
}


bool FileManager::createFolderRecursively(std::string path)
{
    path = convertPath(path);

    int pos = 0;
    int size = path.size();
    bool flag = true;
    while (flag)
    {
        pos++;
        pos = path.find_first_of("\\/", pos);
        //如果返回值为1，或者已经是完整路径了，则不需要创建
        if (pos < 0 || pos >= size)
        {
            flag = createFolder(path.substr(0, pos));
            break;
        }
        flag = createFolder(path.substr(0, pos));

//在win32操作系统中强制
#if (CC_TARGET_PLATFORM == CC_PLATFORM_WIN32)
        flag = true;
#endif
    }

    return flag;
}


void FileManager::deleteFolder(std::string path)
{
    path = convertPath(path);

    // Remove downloaded files
#if (CC_TARGET_PLATFORM != CC_PLATFORM_WIN32)
    string command = "rm -r ";
    // Path may include space.
    command += "\"" + path + "\"";
    system(command.c_str());
#else
    string command = "rd /s /q ";
    // Path may include space.
    command += "\"" + path + "\"";
    system(command.c_str());
#endif
}

bool FileManager::deleteFile(std::string path)
{
    path = convertPath(path);

    if (remove(path.c_str()) != 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

File: lua/io/FileManager.lua
--[[
    对文件进行管理操作
--]]


--[[
    属性
--]]
FileManager =
{
    instance = nil,
}


--[[
    单例类
--]]
FileManager = class("FileManager");


--[[
    获取单例
--]]
function FileManager:getInstance()
    if (not self.instance) then
        local instance = self:new();

        if (instance and instance:init()) then
            self.instance = instance;
        else
            error("创建FileManager单例失败！");

            instance = nil;
        end
    end

    return self.instance;
end


--[[
    初始化
--]]
function FileManager:init()
    self.instance = nil;

    return true;
end


--[[
    删除单例
--]]
function FileManager:destroyInstance()
    self.instance = nil;
end


--[[
    转换地址路径，因为不同平台的分隔符不一样
--]]
function FileManager:convertPath(path)
    if (type(path) ~= "string") then
        error("参数错误！");

        return false;
    end

    if (CCApplication:sharedApplication():getTargetPlatform() == kTargetWindows) then
        return string.gsub(path, "/", "\\");
    else
        return string.gsub(path, "\\", "/");
    end
end


--[[
    检查文件是否存在
--]]
function FileManager:isFileExist(path)
    if (type(path) ~= "string") then
        error("参数错误！");

        return false;
    end

    path = self:convertPath(path);

    -- 检测平台
    -- if (CCApplication:sharedApplication():getTargetPlatform() == kTargetAndroid) then
    --     return CCFileUtils:sharedFileUtils():isFileExist(path);
    -- else
    --     local file, msg = io.open(path, "rb");
    --     if (file) then
    --         file:close();

    --         return true;
    --     else
    --         return false, msg;
    --     end
    -- end

    return CCFileUtils:sharedFileUtils():isFileExist(path);
end


--[[
    给文件等重命名
--]]
function FileManager:reName(originPath, newName)
    if (type(originPath) ~= "string" or type(newName) ~= string) then
        error("参数错误！");

        return false;
    end

    originPath = self:convertPath(originPath);

    --获取父文件夹
    path = self:getFatherFolder(originPath);
    --以防新名字已有文件，先删除
    self:deleteFile(path .. "/" .. newName);

    --执行成功返回0，执行失败返回1
    if (os.execute("rename " .. pathA .. " " .. pathB) == 0) then
        return true;
    else
        return false;
    end
end


--[[
    删除文件
    调用os.remove
    删除成功返回true，删除失败返回nil和出错信息的string
--]]
function FileManager:deleteFile(path)
    if (type(path) ~= "string") then
        error("参数错误！");

        return false;
    end

    path = self:convertPath(path);

    return os.remove(path);
end


--[[
    复制文件
--]]
function FileManager:copyFile(pathA, pathB)
    if (type(pathA) ~= "string" or type(pathB) ~= "string") then
        error("参数错误！");

        return false;
    end

    pathA = self:convertPath(pathA);
    pathB = self:convertPath(pathB);

    --[[ 该方法太复杂，直接使用命令更为快捷
    local fileA, msgA = io.open(pathA, "r");
    if (not fileA) then
        error(msgA);

        return false;
    end

    local fileB, msgB = io.open(pathB, "w");
    if (not fileB) then
        fileA:close();
        error(msgB);

        return false;
    end

    fileB:write(fileA:read("*all"));

    fileA:close();
    fileB:close();

    return true;
    --]]

    --执行成功返回0，执行失败返回1
    if (os.execute("cp " .. pathA .. " " .. pathB) == 0) then
        return true;
    else
        return false;
    end
end


--[[
    新建文件夹
    调用os.execute
    调用成功返回true，失败返回false
    此命令本身可以递归的创建文件夹
--]]
function FileManager:createFolder(path)
    if (type(path) ~= "string") then
        error("参数错误！");

        return false;
    end

    path = self:convertPath(path);

    --执行成功返回0，执行失败返回1
    local flag = nil;
    if (CCApplication:sharedApplication():getTargetPlatform() == kTargetWindows) then
        flag = os.execute("mkdir " .. path);
    else
        flag = os.execute("mkdir -p " .. path);
    end

    if (flag == 0) then
        return true;
    else
        return false;
    end
end


--[[
    算出父文件夹
--]]
function FileManager:getFatherFolder(path)
    if (type(path) ~= "string") then
        error("参数错误！");

        return false;
    end

    path = self:convertPath(path);
    local start, last = nil, nil;

    if (CCApplication:sharedApplication():getTargetPlatform() == kTargetWindows) then
        start, last = findLastOf(path, "\\");
    else
        start, last = findLastOf(path, "/");
    end

    if (start) then
        return string.sub(path, 1, last - 1);
    else
        return "";
    end
end


--[[
    算出文件名，带后缀的
--]]
function FileManager:getFileNameWithSuffix(path)
    if (type(path) ~= "string") then
        error("参数错误！");

        return false;
    end

    path = self:convertPath(path);
    local start, last = nil, nil;

    if (CCApplication:sharedApplication():getTargetPlatform() == kTargetWindows) then
        start, last = findLastOf(path, "\\");
    else
        start, last = findLastOf(path, "/");
    end

    if (start) then
        return string.sub(path, last + 1);
    else
        return path;
    end
end


--[[
    算出文件名，不带后缀的
--]]
function FileManager:getFileName(path)
    if (type(path) ~= "string") then
        error("参数错误！");

        return false;
    end

    path = self:convertPath(path);
    path = self:getFileNameWithSuffix(path);

    local start, last = string.find(path, "%.");
    
    return string.sub(path, 1, start - 1);
end


--[[
    删除文件夹
    调用os.execute
    调用成功返回true，失败返回false
--]]
function FileManager:deleteFolder(path)
    if (type(path) ~= "string") then
        error("参数错误！");

        return false;
    end

    path = self:convertPath(path);

    local command = nil;
    if (CCApplication:sharedApplication():getTargetPlatform() == kTargetWindows) then
        command = "rd /s /q ";
    else
        command = "rm -r ";
    end

    --执行成功返回0，执行失败返回1
    if (os.execute(command .. path) == 0) then
        return true;
    else
        return false;
    end
end


--[[
    复制文件夹
    这个貌似有点问题，别用太多
--]]
function FileManager:copyFolder(pathA, pathB)
    if (type(pathA) ~= "string" or type(pathB) ~= "string") then
        error("参数错误！");

        return false;
    end

    pathA = self:convertPath(pathA);
    pathB = self:convertPath(pathB);

    local command, commandEx = nil, nil;
    if (CCApplication:sharedApplication():getTargetPlatform() == kTargetWindows) then
        command = "robocopy ";
        commandEx = " /E";
    else
        command = "cp -r ";
        commandEx = "";
    end

    --成功返回1，无需复制返回0，路径错误返回16
    if (os.execute(command .. pathA .. " " .. pathB .. commandEx) == 1) then
        return true;
    else
        return false;
    end
end